util/fipstools/inject_hash/macho_parser/macho_parser.c

```diff
@@ -0,0 +1,165 @@
+int read_macho_file(const char *filename, machofile *macho) {
+    FILE *file = NULL;
+    load_cmd *load_commands = NULL;
+    int ret = 1;
+
+    file = fopen(filename, "rb");
+    if (!file) {
+        LOG_ERROR("Error opening file %s", filename);
+        ret = 0;
+        goto end;
+    }
+
+    fread(&macho->macho_header, sizeof(macho_header), 1, file);
+    if(macho->macho_header.magic != MH_MAGIC_64) {
+        LOG_ERROR("File is not a 64-bit Mach-O file");
+        ret = 0;
+        goto end;
+    }
+
+    load_commands = malloc(macho->macho_header.sizeofcmds);
+    fread(load_commands, macho->macho_header.sizeofcmds, 1, file);
+
+    // We're only looking for __text, __const in the __TEXT segment, and the string & symbol tables
+    macho->num_sections = 4;
+    macho->sections = malloc(macho->num_sections * sizeof(section_info));
+
+    uint32_t section_index = 0;
+    for (uint32_t i = 0; i < macho->macho_header.sizeofcmds / BIT_MODIFIER; i += load_commands[i].cmdsize / BIT_MODIFIER) {
+        if (load_commands[i].cmd == LC_SEG) {
+            segment_load_cmd *segment = (segment_load_cmd *)&load_commands[i];
+            if (strcmp(segment->segname, "__TEXT") == 0) {
+                section_data *sections = (section_data *)&segment[1];
+                for (uint32_t j = 0; j < segment->nsects; j++) {
+                    if (strcmp(sections[j].sectname, "__text") == 0 || strcmp(sections[j].sectname, "__const") == 0) {
+                        macho->sections[section_index].offset = sections[j].offset;
+                        macho->sections[section_index].size = sections[j].size;
+                        strcpy(macho->sections[section_index].name, sections[j].sectname);
+                        section_index++;
+                    }
+                }
+            }
+        } else if (load_commands[i].cmd == LC_SYMTAB) {
+            symtab_load_cmd *symtab = (symtab_load_cmd *)&load_commands[i];
+            macho->sections[section_index].offset = symtab->symoff;
+            macho->sections[section_index].size = symtab->nsyms * sizeof(symbol_info);
+            strcpy(macho->sections[section_index].name, "__symbol_table");
+            section_index++;
+            macho->sections[section_index].offset = symtab->stroff;
+            macho->sections[section_index].size = symtab->strsize;
+            strcpy(macho->sections[section_index].name, "__string_table");
+            section_index++;
+        }
+    }
+
+
+end:
+    if (load_commands != NULL) {
+        free(load_commands);
+    }
+    if (file != NULL) {
+        fclose(file);
+    }
+    return ret;
+}
+
+void free_macho_file(machofile *macho) {
+    free(macho->sections);
+    free(macho);
+    macho = NULL;
+}
+
+uint8_t* get_macho_section_data(const char *filename, machofile *macho, const char *section_name, size_t *size, uint32_t *offset) {
+    FILE *file = NULL;
+    uint8_t *section_data = NULL;
```

Comment:
You can move the declaration of `section_data` down to the first place it's assigned to.

Reply:
I'm under the impression that I need to leave it out of the for loop in order to be able to free it, unless I'm misunderstanding you here?

Reply:
You could have it like this:
```
...
    for (uint32_t i = 0; i < macho->num_sections; i++) {
        if (strcmp(macho->sections[i].name, section_name) == 0) {
            uint8_t *section_data = malloc(macho->sections[i].size); // declaration
            if (section_data == NULL) {
                LOG_ERROR("Error allocating memory for section data");
                goto end;
            }

            fseek(file, macho->sections[i].offset, SEEK_SET);
            bytes_read = fread(section_data, 1, macho->sections[i].size, file);
            if (bytes_read != macho->sections[i].size) {
                free(section_data);                                // free on error
                LOG_ERROR("Error reading section data from file %s", filename);
                goto end;
            }
...
```